dissect/evidence/ad1/ad1.py

```diff
@@ -0,0 +1,347 @@
+from __future__ import annotations
+
+import re
+import zlib
+from datetime import datetime, timezone
+from functools import cached_property
+from pathlib import Path, PurePosixPath, PureWindowsPath
+from typing import TYPE_CHECKING, BinaryIO
+
+from dissect.util.stream import AlignedStream, MappingStream, RelativeStream
+
+from dissect.evidence.ad1.c_ad1 import c_ad1
+from dissect.evidence.exceptions import FileNotFoundError, NotADirectoryError, NotASymlinkError
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+
+EntryType = c_ad1.EntryType
+MetaType = c_ad1.MetaType
+FileClassType = c_ad1.FileClassType
+
+
+def atoi(text: str) -> int | str:
+    return int(text) if text.isdigit() else text
+
+
+def natural_keys(text: str | Path) -> list[int | str]:
+    return [atoi(c) for c in re.split(r"(\d+)", str(text))]
+
+
+def find_files(path: Path) -> set[Path]:
+    files = set()
+    for file in path.parent.iterdir():
+        if file.stem == path.stem and re.match(r"^\.ad[0-9]+$", file.suffix.lower()):
+            files.add(file)
+    return sorted(files, key=natural_keys)
+
+
+class AD1:
+    """AccessData Logical Image (AD1v4) implementation.
+
+    Supports ``zlib`` compressed images. Does not directly support encrypted (``b"ADCRYPT"``) images.
+
+    Should be initialized using a list of segment files, e.g.::
+
+        fs = AD1([Path("file.ad1").open("rb"), Path("file.ad2").open("rb")])
+
+    Resources:
+        - Reversing FTK Imager
+        - https://github.com/pcbje/pyad1/blob/master/documentation/AccessData%20Format%20(AD1).asciidoc
+        - https://github.com/al3ks1s/AD1-tools
+        - https://web.archive.org/web/20231013073319/https://tmairi.github.io/posts/dissecting-the-ad1-file-format/
+        - https://al3ks1s.fr/posts/adventures-part-1/
+    """
+
+    def __init__(self, fh: BinaryIO | list[BinaryIO]):
```

Comment:
The typing is currently wrong in `EWF` (maybe you could fix that too? 😇), but I think we'd prefer to support passing in Path objects too. The reason for that is this very well documented PR: https://github.com/fox-it/dissect.evidence/pull/18. 

Kidding aside, with poorly fragmented AD1 files, we could run into the issue that we'd run out of OS file handles if we'd try to open them all simultaneously. For that reason, EWF has a "poor mans LRU" on the segment files. I think it'd be good to implement something similar for AD1.